notebooks/generative/mnist_autoencoder_using_kino.livemd

```diff
@@ -0,0 +1,308 @@
+
+# The noisy image is the input to the network
+# and the original image is the target it's trying to match
+train_data = Stream.zip(batched_noisy_train_images, batched_train_images)
+
+batched_test_images =
+  test_images
+  |> Nx.to_batched(batch_size)
+
+batched_noisy_test_images =
+  test_images
+  |> Nx.to_batched(batch_size)
+  |> Stream.map(add_noise)
+
+test_data = Stream.zip(batched_noisy_test_images, batched_test_images)
+```
+
+Let's see what an element of the input and target look like
+
+```elixir
+{input_batch, target_batch} = Enum.at(train_data, 0)
+{Nx.to_heatmap(input_batch[images: 0]), Nx.to_heatmap(target_batch[images: 0])}
+```
+
+Looks right (and tricky). Let's see how the model does.
+
+```elixir
+params =
+  model
+  |> Axon.Loop.trainer(:mean_squared_error, Axon.Optimizers.adamw(0.001))
+  |> Axon.Loop.validate(model, test_data)
+  |> Axon.Loop.run(train_data, %{}, epochs: 20, compiler: EXLA)
+
+:ok
+```
+
+Now that we have a model that theoretically has learned *something*, we'll see what it's learned by running it on some images from the test set. We'll use Kino to allow us to select the image from the test set to run the model against. To avoid losing the params that took a while to train, we'll create another branch so we can experiment with the params and stop execution when needed without having to retrain.
+
+<!-- livebook:{"branch_parent_index":2} -->
+
+## Evaluation
+
+*A note on branching:*
+
+This section is a branch because "stopping" execution of a cell cause the other cells in the notebook to enter an `Aborted` state and any state created by those cells is lost. This is because stopping execution stops the process that's executing the Livebook cells and holds the corresponing memory.
+
+Since we just spent a bunch of time training the model and don't want to lose that memory state as we continue to experiment, we create a Section "branch". Creating a branch creates a new process with its own memory that's copied from the branch point. Stopping the cell within the branch only stops that the branch process and prevents the notebook from losing the state in the non branched sections.
+
+This does add some memory overhead, but it's worth it so we can experiment without fear.
+
+<!-- livebook:{"break_markdown":true} -->
+
+To use `Kino` to give us an interactive tool to evaluate the model, we'll create a `Kino.Frame` that we can dynamically update. We'll also create a form using `Kino.Control` to allow the user to select which image from the test set they'd like to evaluate the model on. Finally `Kino.Control.stream` enables us to respond to changes in the user's selection when the user clicks the "Render" button.
+
+We can use `Nx.concatenate` to stack the images side by side for a prettier output.
+
+```elixir
+frame = Kino.Frame.new() |> Kino.render()
+
+form =
+  Kino.Control.form(
+    [
+      test_image_index: Kino.Input.number("Test Image Index", default: 0)
+    ],
+    submit: "Render"
+  )
```

Comment:
Idea: we could also just render a "Next" button and use `Kino.Control.stream/3` keeping track of the current index in state and rendering subsequent images on click. This would make the iteration quicker, but if we want to look at specific images, an input makes sense, so whichever you like more!

Reply:
The input seems like a common use case so I'll leave it that way, but maybe I'll use the stateful version in the VAE notebook since it'll be useful to see the steps of the interpolation. Thanks for the pointer!